File: tests/compiler/cfg.test.ts
import { describe, expect, it } from "vitest";

import {
  buildCfgEdges,
  buildCfgSuccessors,
  buildTargetIndexes,
  loopCounterRegister,
} from "../../src/core/passes/cfg.ts";
import type { IrLoopCounter, IrOp, RegisterName } from "../../src/core/types.ts";

const REGISTER_INDEX: Record<RegisterName, number> = {
  "0": 0,
  "1": 1,
  "2": 2,
  "3": 3,
  "4": 4,
  "5": 5,
  "6": 6,
  "7": 7,
  "8": 8,
  "9": 9,
  a: 10,
  b: 11,
  c: 12,
  d: 13,
  e: 14,
};

function plain(opcode: number, mnemonic: string): IrOp {
  return { kind: "plain", opcode, meta: { mnemonic } };
}

function store(register: RegisterName): IrOp {
  return {
    kind: "store",
    register,
    opcode: 0x40 + REGISTER_INDEX[register],
    meta: { mnemonic: `X->П ${register}` },
  };
}

function label(name: string): IrOp {
  return { kind: "label", name };
}

function jump(target: string | number): IrOp {
  return { kind: "jump", target, opcode: 0x51, meta: { mnemonic: "БП" }, targetMeta: {} };
}

function cjump(target: string | number): IrOp {
  return {
    kind: "cjump",
    condition: "==0",
    target,
    opcode: 0x5e,
    meta: { mnemonic: "F x=0" },
    targetMeta: {},
  };
}

function call(target: string | number): IrOp {
  return { kind: "call", target, opcode: 0x53, meta: { mnemonic: "ПП" }, targetMeta: {} };
}

function loop(target: string): IrOp {
  return {
    kind: "loop",
    counter: "L0",
    target,
    opcode: 0x5d,
    meta: { mnemonic: "F L0" },
    targetMeta: {},
  };
}

function halt(): IrOp {
  return { kind: "stop", opcode: 0x50, semantic: "halt", meta: { mnemonic: "С/П" } };
}

function ret(): IrOp {
  return { kind: "return", opcode: 0x52, meta: { mnemonic: "В/О" } };
}

function indirectJump(register: RegisterName, target?: number): IrOp {
  return {
    kind: "indirect-jump",
    register,
    opcode: 0x80 + REGISTER_INDEX[register],
    meta: target === undefined
      ? { mnemonic: `К БП ${register}` }
      : { mnemonic: `К БП ${register}`, comment: `indirect-target=${target}` },
  };
}

function indirectCall(register: RegisterName, target?: number): IrOp {
  return {
    kind: "indirect-call",
    register,
    opcode: 0xa0 + REGISTER_INDEX[register],
    meta: target === undefined
      ? { mnemonic: `К ПП ${register}` }
      : { mnemonic: `К ПП ${register}`, comment: `indirect-target=${target}` },
  };
}

describe("cfg target indexes", () => {
  it("maps labels to op indexes and addresses to executable ops", () => {
    const ops: IrOp[] = [
      plain(0x01, "1"), // address 0
      label("head"), // not addressable
      jump("head"), // address 1, two cells
      plain(0x02, "2"), // address 3
    ];
    const { labelIndex, addressIndex } = buildTargetIndexes(ops);
    expect(labelIndex.get("head")).toBe(1);
    expect(addressIndex.get(0)).toBe(0);
    expect(addressIndex.get(1)).toBe(2);
    expect(addressIndex.get(3)).toBe(3);
    expect(addressIndex.has(2)).toBe(false);
  });
});

describe("cfg successor edges", () => {
  it("links straight-line ops, labels, and stops by fallthrough", () => {
    const ops: IrOp[] = [plain(0x01, "1"), label("x"), store("4"), halt(), plain(0x02, "2")];
    const edges = buildCfgEdges(ops);
    expect(edges[0]).toEqual([{ target: 1, kind: "fallthrough" }]);
    expect(edges[1]).toEqual([{ target: 2, kind: "fallthrough" }]);
    expect(edges[2]).toEqual([{ target: 3, kind: "fallthrough" }]);
    // С/П pauses; pressing С/П resumes from the next cell.
    expect(edges[3]).toEqual([{ target: 4, kind: "fallthrough" }]);
    expect(edges[4]).toEqual([]);
  });

  it("gives jumps a single jump edge and conditionals jump plus fallthrough", () => {
    const ops: IrOp[] = [cjump("done"), plain(0x01, "1"), jump("done"), label("done"), halt()];
    const edges = buildCfgEdges(ops);
    expect(edges[0]).toEqual([
      { target: 3, kind: "jump" },
      { target: 1, kind: "fallthrough" },
    ]);
    expect(edges[2]).toEqual([{ target: 3, kind: "jump" }]);
  });

  it("treats counted loops like conditionals", () => {
    const ops: IrOp[] = [label("head"), plain(0x01, "1"), loop("head"), halt()];
    const edges = buildCfgEdges(ops);
    expect(edges[2]).toEqual([
      { target: 0, kind: "jump" },
      { target: 3, kind: "fallthrough" },
    ]);
  });

  it("resolves numeric targets through machine addresses", () => {
    const ops: IrOp[] = [
      jump(3), // address 0, two cells
      plain(0x01, "1"), // address 2
      plain(0x02, "2"), // address 3
    ];
    const edges = buildCfgEdges(ops);
    expect(edges[0]).toEqual([{ target: 2, kind: "jump" }]);
  });

  it("wires returns to every call continuation instead of call fallthrough", () => {
    const ops: IrOp[] = [
      call("helper"), // continuation index 1
      halt(),
      label("helper"),
      plain(0x01, "1"),
      ret(),
    ];
    const edges = buildCfgEdges(ops);
    expect(edges[0]).toEqual([{ target: 2, kind: "jump" }]);
    expect(edges[4]).toEqual([{ target: 1, kind: "normal" }]);
  });

  it("uses proved indirect targets and leaves unknown indirect flow without edges", () => {
    const ops: IrOp[] = [
      indirectJump("7", 3), // address 0
      indirectJump("8"), // address 1, unknown target
      plain(0x01, "1"), // address 2
      plain(0x02, "2"), // address 3
    ];
    const edges = buildCfgEdges(ops);
    expect(edges[0]).toEqual([{ target: 3, kind: "jump" }]);
    expect(edges[1]).toEqual([]);
  });

  it("adds indirect-call fallthrough only when requested", () => {
    const ops: IrOp[] = [
      indirectCall("7", 2), // address 0
      halt(), // address 1
      plain(0x01, "1"), // address 2
      ret(),
    ];
    const withoutFallthrough = buildCfgEdges(ops);
    expect(withoutFallthrough[0]).toEqual([{ target: 2, kind: "jump" }]);
    // Return still reaches the call continuation.
    expect(withoutFallthrough[3]).toEqual([{ target: 1, kind: "normal" }]);

    const withFallthrough = buildCfgEdges(ops, { indirectCallFallthrough: true });
    expect(withFallthrough[0]).toEqual([
      { target: 2, kind: "jump" },
      { target: 1, kind: "fallthrough" },
    ]);
  });

  it("projects plain successor indexes for consumers that ignore edge kinds", () => {
    const ops: IrOp[] = [cjump("done"), plain(0x01, "1"), label("done"), halt()];
    expect(buildCfgSuccessors(ops)).toEqual([[2, 1], [2], [3], []]);
  });
});

describe("loopCounterRegister", () => {
  it("maps F L0..F L3 to memory registers 0..3", () => {
    const expected: Record<IrLoopCounter, RegisterName> = {
      L0: "0",
      L1: "1",
      L2: "2",
      L3: "3",
    };
    for (const [counter, register] of Object.entries(expected)) {
      expect(loopCounterRegister(counter as IrLoopCounter)).toBe(register);
    }
  });
});

File: native/tests/cfg_test.cpp
#include "mkpro/core/passes/cfg.hpp"

#include "test_support.hpp"

#include <optional>
#include <string>
#include <utility>
#include <vector>

namespace mkpro::tests {

namespace {

IrOp plain(int opcode, std::string mnemonic) {
  IrOp op;
  op.kind = IrKind::Plain;
  op.opcode = opcode;
  op.meta.mnemonic = std::move(mnemonic);
  return op;
}

IrOp store(std::string register_name) {
  IrOp op;
  op.kind = IrKind::Store;
  op.register_name = std::move(register_name);
  op.opcode = 0x40;
  op.meta.mnemonic = "X->П " + op.register_name;
  return op;
}

IrOp label(std::string name) {
  IrOp op;
  op.kind = IrKind::Label;
  op.name = std::move(name);
  return op;
}

IrOp jump(IrTarget target) {
  IrOp op;
  op.kind = IrKind::Jump;
  op.target = std::move(target);
  op.opcode = 0x51;
  op.meta.mnemonic = "БП";
  return op;
}

IrOp cjump(IrTarget target) {
  IrOp op;
  op.kind = IrKind::CondJump;
  op.condition = "==0";
  op.target = std::move(target);
  op.opcode = 0x5e;
  op.meta.mnemonic = "F x=0";
  return op;
}

IrOp call(IrTarget target) {
  IrOp op;
  op.kind = IrKind::Call;
  op.target = std::move(target);
  op.opcode = 0x53;
  op.meta.mnemonic = "ПП";
  return op;
}

IrOp loop(std::string target) {
  IrOp op;
  op.kind = IrKind::Loop;
  op.counter = "L0";
  op.target = std::move(target);
  op.opcode = 0x5d;
  op.meta.mnemonic = "F L0";
  return op;
}

IrOp halt() {
  IrOp op;
  op.kind = IrKind::Stop;
  op.opcode = 0x50;
  op.semantic = "halt";
  op.meta.mnemonic = "С/П";
  return op;
}

IrOp ret() {
  IrOp op;
  op.kind = IrKind::Return;
  op.opcode = 0x52;
  op.meta.mnemonic = "В/О";
  return op;
}

IrOp indirect_jump(std::string register_name, std::optional<int> target = std::nullopt) {
  IrOp op;
  op.kind = IrKind::IndirectJump;
  op.register_name = std::move(register_name);
  op.opcode = 0x80;
  op.meta.mnemonic = "К БП " + op.register_name;
  if (target.has_value())
    op.meta.comment = "indirect-target=" + std::to_string(*target);
  return op;
}

IrOp indirect_call(std::string register_name, std::optional<int> target = std::nullopt) {
  IrOp op;
  op.kind = IrKind::IndirectCall;
  op.register_name = std::move(register_name);
  op.opcode = 0xa0;
  op.meta.mnemonic = "К ПП " + op.register_name;
  if (target.has_value())
    op.meta.comment = "indirect-target=" + std::to_string(*target);
  return op;
}

bool edges_equal(const std::vector<core::passes::CfgEdge>& actual,
                 const std::vector<std::pair<int, core::passes::CfgEdgeKind>>& expected) {
  if (actual.size() != expected.size())
    return false;
  for (std::size_t index = 0; index < actual.size(); ++index) {
    if (actual.at(index).target != expected.at(index).first ||
        actual.at(index).kind != expected.at(index).second) {
      return false;
    }
  }
  return true;
}

void require_edges(const std::vector<std::vector<core::passes::CfgEdge>>& edges, std::size_t index,
                   const std::vector<std::pair<int, core::passes::CfgEdgeKind>>& expected,
                   const std::string& context) {
  require(index < edges.size(), context + " edge index exists");
  require(edges_equal(edges.at(index), expected), context + " edges should match TypeScript");
}

} // namespace

void cfg_matches_typescript_contract() {
  using core::passes::build_cfg_edges;
  using core::passes::build_cfg_successors;
  using core::passes::build_control_flow_graph;
  using core::passes::build_target_indexes;
  using core::passes::BuildCfgOptions;
  using core::passes::CfgEdgeKind;
  using core::passes::CfgUncertaintyKind;
  using core::passes::loop_counter_register;

  {
    const std::vector<IrOp> ops = {
        plain(0x01, "1"),
        label("head"),
        jump(std::string("head")),
        plain(0x02, "2"),
    };
    const core::passes::CfgTargetIndexes indexes = build_target_indexes(ops);
    require(indexes.label_index.at("head") == 1, "cfg labels should map to op indexes");
    require(indexes.address_index.at(0) == 0, "cfg address 0 should map to first op");
    require(indexes.address_index.at(1) == 2, "cfg address 1 should map to jump op");
    require(indexes.address_index.at(3) == 3, "cfg address 3 should map to last op");
    require(!indexes.address_index.contains(2), "cfg should not map address operands as op starts");
  }

  {
    const std::vector<IrOp> ops = {plain(0x01, "1"), label("x"), store("4"), halt(),
                                   plain(0x02, "2")};
    const std::vector<std::vector<core::passes::CfgEdge>> edges = build_cfg_edges(ops);
    require_edges(edges, 0, {{1, CfgEdgeKind::Fallthrough}}, "plain fallthrough");
    require_edges(edges, 1, {{2, CfgEdgeKind::Fallthrough}}, "label fallthrough");
    require_edges(edges, 2, {{3, CfgEdgeKind::Fallthrough}}, "store fallthrough");
    require_edges(edges, 3, {{4, CfgEdgeKind::Fallthrough}}, "stop resume fallthrough");
    require_edges(edges, 4, {}, "terminal plain");
  }

  {
    const std::vector<IrOp> ops = {cjump(std::string("done")), plain(0x01, "1"),
                                   jump(std::string("done")), label("done"), halt()};
    const std::vector<std::vector<core::passes::CfgEdge>> edges = build_cfg_edges(ops);
    require_edges(edges, 0, {{3, CfgEdgeKind::Jump}, {1, CfgEdgeKind::Fallthrough}},
                  "conditional jump");
    require_edges(edges, 2, {{3, CfgEdgeKind::Jump}}, "direct jump");
  }

  {
    const std::vector<IrOp> ops = {label("head"), plain(0x01, "1"), loop("head"), halt()};
    require_edges(build_cfg_edges(ops), 2, {{0, CfgEdgeKind::Jump}, {3, CfgEdgeKind::Fallthrough}},
                  "counted loop");
  }

  {
    const std::vector<IrOp> ops = {jump(3), plain(0x01, "1"), plain(0x02, "2")};
    require_edges(build_cfg_edges(ops), 0, {{2, CfgEdgeKind::Jump}}, "numeric target");
  }

  {
    const std::vector<IrOp> ops = {call(std::string("helper")), halt(), label("helper"),
                                   plain(0x01, "1"), ret()};
    const std::vector<std::vector<core::passes::CfgEdge>> edges = build_cfg_edges(ops);
    require_edges(edges, 0, {{2, CfgEdgeKind::Jump}}, "direct call target");
    require_edges(edges, 4, {{1, CfgEdgeKind::Normal}}, "direct return continuation");
  }

  {
    const std::vector<IrOp> ops = {indirect_jump("7", 3), indirect_jump("8"), plain(0x01, "1"),
                                   plain(0x02, "2")};
    const core::passes::ControlFlowGraph graph = build_control_flow_graph(ops);
    require_edges(graph.edges, 0, {{3, CfgEdgeKind::Jump}}, "proved indirect jump");
    require_edges(graph.edges, 1,
                  {{0, CfgEdgeKind::Jump},
                   {1, CfgEdgeKind::Jump},
                   {2, CfgEdgeKind::Jump},
                   {3, CfgEdgeKind::Jump}},
                  "conservative unknown indirect jump");
    require(graph.uncertainties.size() == 1U && graph.uncertainties.front().source == 1 &&
                graph.uncertainties.front().kind == CfgUncertaintyKind::UnknownIndirectTarget,
            "cfg did not expose the unknown indirect target proof gap");

    const std::vector<std::vector<core::passes::CfgEdge>> sparse =
        build_cfg_edges(ops, BuildCfgOptions{.unknown_indirect_flow_to_all = false});
    require_edges(sparse, 1, {}, "explicit sparse unknown indirect jump");
  }

  {
    IrOp dispatch = indirect_jump("7");
    dispatch.meta.indirect_flow_targets = std::vector<IrTarget>{IrTarget{1}, IrTarget{2}};
    const std::vector<IrOp> ops = {dispatch, plain(0x01, "1"), plain(0x02, "2")};
    const core::passes::ControlFlowGraph graph = build_control_flow_graph(ops);

    require_edges(graph.edges, 0, {{1, CfgEdgeKind::Jump}, {2, CfgEdgeKind::Jump}},
                  "typed multi-target indirect jump");
    require(graph.targets_are_exact(),
            "resolved typed multi-target indirect jump was marked uncertain");
  }

  {
    const std::vector<IrOp> ops = {jump(std::string("missing")), plain(0x01, "1")};
    const core::passes::ControlFlowGraph graph = build_control_flow_graph(ops);

    require_edges(graph.edges, 0, {{0, CfgEdgeKind::Jump}, {1, CfgEdgeKind::Jump}},
                  "conservative unresolved direct jump");
    require(graph.uncertainties.size() == 1U &&
                graph.uncertainties.front().kind == CfgUncertaintyKind::UnresolvedDirectTarget,
            "cfg did not expose the unresolved direct target proof gap");
  }

  {
    const std::vector<IrOp> ops = {indirect_call("7", 2), halt(), plain(0x01, "1"), ret()};
    const std::vector<std::vector<core::passes::CfgEdge>> without = build_cfg_edges(ops);
    require_edges(without, 0, {{2, CfgEdgeKind::Jump}}, "indirect call without fallthrough");
    require_edges(without, 3, {{1, CfgEdgeKind::Normal}}, "indirect return continuation");

    const std::vector<std::vector<core::passes::CfgEdge>> with =
        build_cfg_edges(ops, BuildCfgOptions{.indirect_call_fallthrough = true});
    require_edges(with, 0, {{2, CfgEdgeKind::Jump}, {1, CfgEdgeKind::Fallthrough}},
                  "indirect call with fallthrough");
  }

  {
    const std::vector<IrOp> ops = {cjump(std::string("done")), plain(0x01, "1"), label("done"),
                                   halt()};
    require(build_cfg_successors(ops) == std::vector<std::vector<int>>{{2, 1}, {2}, {3}, {}},
            "cfg plain successors should drop edge kinds");
  }

  require(loop_counter_register("L0") == "0", "L0 should map to R0");
  require(loop_counter_register("L1") == "1", "L1 should map to R1");
  require(loop_counter_register("L2") == "2", "L2 should map to R2");
  require(loop_counter_register("L3") == "3", "L3 should map to R3");
}

} // namespace mkpro::tests
